### ISSUE TITLE
Deadlocks when hot reloading configs

### ISSUE THREAD

**Issue by ramirezmike** [TimUntersberger/nog#228]
**Version (which release did you use?)**
development branch a0bb9eedcd6a7f94781246a1697acb47c07e387e

**Describe the bug**

There are at least two distinct deadlocks that can occasionally occur while changing modes or triggering hot-reloading of the config

**To Reproduce**

**Deadlock A:**
This deadlock happens while switching modes. I believe what is occurring is that the code to switch modes locks the **AppState** object and the **inner** object of the **KBManager** while nog's AppBar at the top of the screen is also doing so in order to redraw components.

My components config in the bar looks like this

    components: #{
        left: [components.workspaces()],
        center: [components.current_window()],
        right: [
            components.active_mode(),
            components.padding(5),
            components.split_direction("↕️", "↔️"),
            components.padding(1),
            components.date("%e %b"),
            components.padding(1),
            components.time("%T"),
            components.padding(1)
        ]
    }

And I have two monitors. I think the # and type of components * the number of monitors has an effect on how likely this deadlock occurs (plus things like cpu load and what programs you have running, I imagine)

**(Suspected) Deadlock B:**
This one occurs when hot-reloading the config, which I usually end up causing by my toggle_config bindings

nog.bind("Alt+I", () => nog.toggle_config("display_app_bar"))
nog.bind("Alt+Shift+I", () => nog.toggle_config("remove_task_bar"))

_Occasionally_ when I trigger this (or when an app takes up the fullscreen and triggers nog to hide the appbar) nog just appears to stop working. From the logs the last thing I see is that it unregisters all keybindings, but nog also doesn't respond to clicks to change workspace or manages new programs.

**Proposed Solution for Deadlock A**
I made a draft pull request [here](https://github.com/TimUntersberger/nog/pull/227) that is one approach to resolving deadlock A. Currently, if a component makes use of AppState it will lock the state to pull information. One of them, the ActiveMode component, locks the state _and_ makes a call to get the current mode which locks the inner of the KBManager. I suspect this is one half of the deadlock. 

I figured that since the components are primarily about rendering information to the screen and refresh milliseconds later, it's OK to let them fail occasionally, so I updated all the components' state locks to instead try to lock for 20 ms and if not then just return an empty string (and also introduced a try_get_mode to the KBManager for the same reason). This would mean that while a component can't grab a state it will render a "frame" of blank text, but...

A) this likely will only happen while the user is doing something to change the state of nog anyway (like changing the modes)
B) it will likely be able to lock the state the next time it attempts

Another approach would be to rewrite the components to return an error when they can't lock and then if an error occurs, just have the component not clear what it's currently displaying. (Theoretically, I guess it's possible that every time a component attempts to lock it fails and then nog never updates that component's display, but that seems incredibly unlikely 🤔)

**Research on Deadlock B**
I have a hunch that this is related to the Windows [SendMessageA](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-sendmessagea) function. This happens when trying to close the app_bar; it calls SendMessageA on the app_bar window which "does not return until the window procedure has processed the message."

What I've seen happen is that closing the app_bar locks the state, then it trigger this SendMessageA which fires off a window close event. Meanwhile, a separate thread is attempting to draw the app_bar and is attempting to lock the state repeatedly (WindowEvent::Draw in twm/src/bar/create.rs). So, it deadlocks with the state already locked in the SendMessageA call and the handler waiting on the Draw event to finish before handling the SendMessageA's close event and the Draw event waiting on the state to be unlocked. I suspect our deadlock detection won't pick this up because the lock is being caused by SendMessageA stopping a thread.

I was able to prevent this deadlock by using PostMessageA instead which queues the request rather than waiting on it, but on my machine when I tried that it introduced flashing components on the appbar. Haven't had time to dig into this further, unfortunately.

**Comment by ramirezmike** [comment]
I re-implemented an approach for deadlock B using PostMessageA that isn't causing the flashing components I was seeing before (which makes sense since this should only be called when the appbar gets closed). Both fixes are in pull request #227

**Comment by keepitsane** [comment]
I am currently running the latest deadlock branch and am still experiencing the `Deadlock B` case. My logs seem to match your description as well.

**Comment by ramirezmike** [comment]
@keepitsane if possible, could you post some of the end of your logs?

how often have you run into it?
how long had nog been running when it freezes up?
when it happens, is it triggered by toggling/updating the config via a keybinding? If so, what is the binding and if not what action are you taking to cause it?

also, just to make sure we're ruling it out: when it happens does everything on the app bar freeze including the time ticking?

**Comment by keepitsane** [comment]
> how often have you run into it?
Basically daily anytime I have to to toggle the taskbar to access the tray, volume, etc
> how long had nog been running when it freezes up?
I have add it freeze up within seconds of running and some after hours.

> when it happens, is it triggered by toggling/updating the config via a keybinding? If so, what is the binding and if not what action are you taking to cause it?
I only toggle my taskbar using the keybinding and my binding is `ALT+CONTROL+I`

> also, just to make sure we're ruling it out: when it happens does everything on the app bar freeze including the time ticking?
I have had two kinds of deadlocks. The first where it freezes and the appbar stays and the clock freezes as well. The second is when nog is still running, but the appbar disappears. Task manager shows that it is still running and if I open up a new instance of Nog the system tray shows two Nog icons.

Here is ending bit of my log file when the deadlock happened (note I was able to toggle the taskbar a few times successfully, before the deadlock happened):
```
[2021-02-18 19:16:08.624749 -05:00] DEBUG [twm\src\main.rs:934] Received keybinding Keybinding(ALT+CONTROL+I, 5, false, 3073, None)
[2021-02-18 19:16:08.639442 -05:00] DEBUG [twm\src\keybindings.rs:234] KbManager received UnregisterKeybindings
[2021-02-18 19:16:08.641942 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+I, 0, false, 1073, None)
[2021-02-18 19:16:08.641942 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Q, 1, false, 1081, None)
[2021-02-18 19:16:08.641942 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+M, 2, false, 1077, None)
[2021-02-18 19:16:08.641942 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+X, 3, false, 1088, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+R, 4, false, 1082, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+I, 5, false, 3073, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+L, 6, false, 1076, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+H, 7, false, 1072, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+K, 8, false, 1075, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+J, 9, false, 1074, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+L, 10, false, 5076, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+H, 11, false, 5072, None)
[2021-02-18 19:16:08.642446 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+K, 12, false, 5075, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+J, 13, false, 5074, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Plus, 14, false, 1187, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Minus, 15, false, 1189, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(CONTROL+SHIFT+F, 16, false, 6070, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+W, 17, true, 3087, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+F, 18, false, 1070, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Zero, 19, false, 5048, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+One, 20, false, 5049, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Two, 21, false, 5050, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Three, 22, false, 5051, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Four, 23, false, 5052, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Five, 24, false, 5053, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Six, 25, false, 5054, None)
[2021-02-18 19:16:08.642944 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Seven, 26, false, 5055, None)
[2021-02-18 19:16:08.643441 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Eight, 27, false, 5056, None)
[2021-02-18 19:16:08.643441 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+SHIFT+Nine, 28, false, 5057, None)
[2021-02-18 19:16:08.643441 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Zero, 29, false, 3048, None)
[2021-02-18 19:16:08.643441 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+One, 30, false, 3049, None)
[2021-02-18 19:16:08.643441 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Two, 31, false, 3050, None)
[2021-02-18 19:16:08.643441 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Three, 32, false, 3051, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Four, 33, false, 3052, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Five, 34, false, 3053, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Six, 35, false, 3054, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Seven, 36, false, 3055, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Eight, 37, false, 3056, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+CONTROL+Nine, 38, false, 3057, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Zero, 39, false, 1048, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+One, 40, false, 1049, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Two, 41, false, 1050, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Three, 42, false, 1051, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Four, 43, false, 1052, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Five, 44, false, 1053, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Six, 45, false, 1054, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Seven, 46, false, 1055, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Eight, 47, false, 1056, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+Nine, 48, false, 1057, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+B, 49, false, 1066, None)
[2021-02-18 19:16:08.643943 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+C, 50, false, 1067, None)
[2021-02-18 19:16:08.644443 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+E, 51, false, 1069, None)
[2021-02-18 19:16:08.644443 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+S, 52, false, 1083, None)
[2021-02-18 19:16:08.644443 -05:00] INFO [twm\src\keybindings.rs:67] Unregistering Keybinding(ALT+V, 53, false, 1086, None)
```

**Comment by ramirezmike** [comment]
@keepitsane ok, sorry got more questions

I know you said you were using the latest deadlock branch, just want to double-check for my own sanity you're using the build from [here](https://github.com/TimUntersberger/nog/releases/tag/deadlocks-release) (the newest version was 3 days ago)

how many monitors do you use?

do you run nog in multi monitor mode?

how many workspaces are you using?

what is your component setup on your app bar(s)?

do you use modes? if so, what is your setup and how often do you use them?

I'll probably come back with more questions. I've been using the same build all week and haven't run into these yet when I kept hitting them before.

**Comment by TimUntersberger** [comment]
> he second is when nog is still running, but the appbar disappears. Task manager shows that it is still running and if I open up a new instance of Nog the system tray shows two Nog icons.

I had this happen too. You should be able to retoggle work mode to make the appbar reappear.

**Comment by keepitsane** [comment]
> I know you said you were using the latest deadlock branch, just want to double-check for my own sanity you're using the build from [here](https://github.com/TimUntersberger/nog/releases/tag/deadlocks-release) (the newest version was 3 days ago)


@ramirezmike Well this is embarrassing, apparently I wasn't using the deadlock branch. Normally when I run a fresh build from lets say my downloads folder, Nog always used to make a copy of itself and update the exe located in the folder with the config files in the Roaming directory. This way my shortcut and auto start always use the latest version and I can delete the copy in my downloads folder. Seems like this functionality is broken for all branches.

I'll continue to properly test out this branch and see if I encounter the deadlocks, so far I can quickly toggle back and forth and the issue seems to be resolved. I am really sorry about the confusion this probably caused you!

> @keepitsane ok, sorry got more questions


Please feel free to ask as many questions as you would like! Nog (or is it called TWM now bit confused about that) has become an essential part of my daily workflow so I am here to help as much as possible. 😄

**Comment by TimUntersberger** [comment]
> or is it called TWM now bit confused about that

It is still called `nog`. Looks like I forgot to rename the executable

**Comment by ramirezmike** [comment]
> I am really sorry about the confusion this probably caused you!

no worries, glad we pieced that together.

>Nog (or is it called TWM now bit confused about that) has become an essential part of my daily workflow so I am here to help as much as possible. 

same 👍

>I'll continue to properly test out this branch and see if I encounter the deadlocks, so far I can quickly toggle back and forth and the issue seems to be resolved.

@TimUntersberger I think once we get the @keepitsane seal-of-approval we can merge the deadlocks branch into master and then make the new stable release.

**Comment by TimUntersberger** [comment]
>@TimUntersberger I think once we get the @keepitsane seal-of-approval we can merge the deadlocks branch into master and then make the new stable release.

👍

**Comment by keepitsane** [comment]
So far deadlock branch has been smooth for me. Great work! 

I think the TWM name issue should be changed back before the new stable release, but seems this is branch is safe to be merged into master.

**Comment by ramirezmike** [comment]
Cool cool, merged deadlocks branch so closing this issue.
